src/internal/connector/exchange/service_restore.go

```diff
@@ -522,6 +522,15 @@ func establishMailRestoreLocation(
 	folderID := rootFolderAlias
 	pb := path.Builder{}
 
+	// Only populate the cache if we actually had to create it. Since we set
+	// newCache to false in this we'll only try to populate it once per function
+	// call even if we make a new cache.
+	if isNewCache {
+		if err := mfc.Populate(ctx, folderID, ""); err != nil {
```

Comment:
Just a minor note, feel free to override me on this. This will cause the system to fetch all folders for mail (and the other types) as it's using the default root folder. This is fine, but not quite as efficient if all we really want is the folders from the top-level restore folder down

However, if the root folder isn't used then there's a bit of a chicken and egg problem. That's because the folder ID for the root of the cache is required to populate the cache. However, we won't know the ID of the top-level restore folder until we either create it or check if it exists

Again, this is just a question of efficiency, so it can be ignored if that's best in the long-run. I think this solution is probably a bit better suited for the case where we allow restoring to an existing folder though

Reply:
I had a similar thought. This would enable `RestoreToExistingLocation` which in essence is anywhere in the Ecosystem. We would be creating a larger lookup table as a tradeoff. @ryanfkeepers Do you have any feelings on this?

Reply:
As a sort of middle ground, is there a way we could restrict that population to only a certain category of folder?  "populate, starting from this folder (the root), only mail|contact|event-system folders"?

Reply:
I wouldn't call that a middle ground. The search trees generated by the populate function are dynamic.
In the case of Restore, the original intention was that `Corso` would restore to a "new" folder at the beginning of every restore operation. The previous optimization was to use a `do-while` style of approach that ensured that `Corso` would not find the folder within the cache as it was not already populated. 

From that space, the Restore folder would be used as the base of the  Search tree. All following queries would search only the "restore portion" of the search space. For Contacts, the restore folder is created the first time within the cache and all following restores are sent to that one folder. 

As far as population, in general, Events already ignores the rootID and base due to its unique structure.